### ISSUE TITLE
print_queries.x does not work for prepbufr subset ASCATW

### ISSUE THREAD

**Issue by YoulongXia-NOAA** [JCSDA-internal/ioda-converters#1032]
## Current behavior (describe the bug)

When I work for regional prepbufr data rap.t06z.prepbufr.tm00 (for RRFS system) on 2022082406, I used split_by_subset.x to subset for  ADPSFC, ADPUPA, AIRCAR, AIRCFT, ASCATW, GPSIPW , MSONET, PROFLR, RASSDA, SATWND, SFCSHP, and VADWND.  print_queries.x works for all subsets except for ASCATW. When I execute "print_queries.x  ASCATW", it has error below:

Available subsets:
ASCATW
Total number of subsets found: 1

terminate called without an active exception
Abort

The data is in /scratch1/NCEPDEV/da/Youlong.Xia/test/bufrCloud/testinput/regional/rrfs . I also tested for ASCATW sub-set gdas.t12z.prepbufr in /scratch1/NCEPDEV/da/Youlong.Xia/test/bufrCloud/testinput/preBufr.  Both datasets are on **Hera**.  This also affects bufr2ioda.x to produce ioda-format data when the whole prepbufr file is used (for both global GDAS and regional RRFS).
 
## To Reproduce

> Hera machine is used

> source modulefile.iodaconv.query.hera.csh in /scratch1/NCEPDEV/da/Youlong.Xia/test

> Steps to reproduce the behavior

1.  In /scratch1/NCEPDEV/da/Youlong.Xia/test/bufrCloud/testinput/regional/rrfs, do "../../preBufr/print_queries.x ASCATW"
2. For yaml file, in /scratch1/NCEPDEV/da/Youlong.Xia/test/bufrCloud/yaml/regional, do "../type_bufr2ioda.x bufr_ncep_prepbufr_cloud.yaml", the error will be appeared. If you replace ASCATW with the other subsets' name, it works without errors. 
3.  A more simple example is testWithStar.yaml or testWithHEADR.yaml. The first has error as the above. The second works with warning message but all values for latitude, longitude, and stationIdentification are missing with"-". I used debufr to output text filr and checked, these values exists. This means that even I use HEADR, it still does not work correctly. If you replace ASCATW in yaml file, and then check ioda output, it works corretcly.
...

## Expected behavior

The print_queries.x should work correctly as it works for the other subsets.

**Comment by rmclaren** [comment]
I just pushed a fix for some bugs including this one. Please update query_cxx and try it again. If it works please don't forget to close this ticket.

**Comment by YoulongXia-NOAA** [comment]
@rmclaren, I am using query_cxx_types, do I need to shift to query_cxx or you also update for types so that I can update types and use query_cxx_types.

**Comment by rmclaren** [comment]
@YoulongXia-NOAA I just updated query_cxx_types so please give it a try.

**Comment by YoulongXia-NOAA** [comment]
@rmclaren, when I tried to build  feature/query_cxx_types, I have the following error:
[ 13%] Building Fortran object oops/src/CMakeFiles/oops.dir/oops/generic/fft_spe  ctral2gpoint_f.F90.o
[ 13%] Building Fortran object oops/src/CMakeFiles/oops.dir/oops/generic/fft_gpo  int2spectral_f.F90.o
[ 13%] Linking CXX shared library ../../lib/liboops.so
[ 13%] Built target oops
make: *** [all] Error 2

I completely removed my buildType directory and do such a bulid two times, it has the same error. For feature/query_cxx, I built it successfully. Now my question is if I need to re-build NCEPbufr library. If yes, should "develop" or "type" branch be used? My work is on Hera (/scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle).

**Comment by YoulongXia-NOAA** [comment]
@rmclaren, feature/query_enhancements rather than "type for NCEPLIBS-bufr.

**Comment by rmclaren** [comment]
@YoulongXia-NOAA I don't see an error and also that output is from oops... Maybe try clearing out your build directory and trying again from scratch?

**Comment by rmclaren** [comment]
@YoulongXia-NOAA Try copying all the output for your build. The fragment you have does not show were the error happened.

**Comment by YoulongXia-NOAA** [comment]
@rmclaren, now I did the following works:
(1) I  did "git pull" for feature/query_enhancements of NCEPLIBS-bufr and re-built /scratch1/NCEPDEV/da/Youlong.Xia/NCEPLIBS-bufr/buildE
(2) I open a new window and source buildE_modulefile.iodaconv.query.hera.csh in/scratch1/NCEPDEV/da/Youlong.Xia/test (using newly built NCEPLIBS-bufr library) 
(3) I completely removed buildType, did ecbuild and the log file is ecbuild.log in /scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle/buildType
(4) make -j8 > make_output.txt, and make_output.txt is in /scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle/buildType

As usual, I git pull all sub-directories in /scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle to ensure that they are updated. I am wondering why the query_cxx works but query_cxx_types does not on Hera.

**Comment by YoulongXia-NOAA** [comment]
@rmclaren, the following messages seem to be related with the bufr: 
/scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider.cpp: In member function 'Ingester::bufr::TypeInfo Ingester::bufr::DataProvider::getTypeInfo(Ingester::bufr::FortranIdx) const':
/scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider.cpp:116:9: error: 'nemdefs_f' was not declared in this scope
  116 |         nemdefs_f(fileUnit_,
      |         ^~~~~~~~~
/scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider.cpp:142:13: error: 'nemtab_f' was not declared in this scope
  142 |             nemtab_f(bufrLoc_,
      |             ^~~~~~~~
/scratch1/NCEPDEV/da/Youlong.Xia/test/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider.cpp:148:13: error: 'nemtbb_f' was not declared in this scope
  148 |             nemtbb_f(bufrLoc_,
      |             ^~~~~~~~
make[2]: *** [iodaconv/tools/bufr/CMakeFiles/print_queries.x.dir/__/__/src/bufr/BufrParser/Query/DataProvider.cpp.o] Error 1
make[1]: *** [iodaconv/tools/bufr/CMakeFiles/print_queries.x.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2

**Comment by rmclaren** [comment]
You may need to update NCEPLIB-bufr in your environment. nemdefs and nemtab are recently added. NCEPLIB-bufr develop should have these changes.

**Comment by rmclaren** [comment]
Please use the develop version of NCEPLIB_bufr. The query_enhancements branch has been merged, so that is a dead branch.

**Comment by YoulongXia-NOAA** [comment]
@rmclaren, you are correct. When used  develop version of NCEPLIB_bufr, ioda-bundle is built. I have removed this dead branch from my local repository to avoid the future misuse.  I also tested print_queries.x and bufr2ioda.x, both work now for ASCATW case. I will double check with Praveen and see if this fix works for his case. If yes, I will close this issue.

**Comment by YoulongXia-NOAA** [comment]
@KumarPraven, when your time is available, could you please test Ron's fix and see if it works your case as well? If yes, please let me know so that I am able to close this issue.

**Comment by YoulongXia-NOAA** [comment]
@rmclaren, thank you for your fixing. Nice job!
